ext/rugged/rugged_blob.c

```diff
@@ -300,18 +275,34 @@ static VALUE rb_git_blob_from_io(int argc, VALUE *argv, VALUE klass)
 		hint_path = StringValueCStr(rb_hint_path);
 	}
 
-	payload.exception = 0;
-	payload.rb_data = rb_io;
+	error = git_blob_create_fromstream(&stream, repo, hint_path);
+	if (error)
+		goto cleanup;
+
+	rb_read_args[0] = rb_io;
+	rb_read_args[1] = INT2FIX(max_length);
+
+	do {
+		rb_buffer = rb_protect(rb_read_check, (VALUE)rb_read_args, &exception);
+
+		if (exception)
+			goto cleanup;
+
+		if (NIL_P(rb_buffer))
```

Comment:
You need a `TYPE(rb_buffer)` check here. Otherwise this will segfault if our callback returns something that is not a string.